contracts/peg/BasePSM.sol

```diff
@@ -0,0 +1,229 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.4;
+
+import {IPCVDeposit, PCVDeposit} from "./../pcv/PCVDeposit.sol";
+import {IBasePSM} from "./IBasePSM.sol";
+import {OracleRef, Decimal, SafeCast} from "./../refs/OracleRef.sol";
+import {Constants} from "../Constants.sol";
+import {SafeERC20, IERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {Math} from "@openzeppelin/contracts/utils/math/Math.sol";
+import {TribeRoles} from "../core/TribeRoles.sol";
+
+abstract contract BasePSM is IBasePSM, OracleRef, PCVDeposit {
+    using Decimal for Decimal.D256;
+    using SafeCast for *;
+    using SafeERC20 for IERC20;
+
+    /// @notice the token this PSM will exchange for VOLT
+    IERC20 public immutable override underlyingToken;
+
+    /// @notice constructor
+    /// @param params PSM constructor parameter struct
+    constructor(OracleParams memory params, IERC20 _underlyingToken)
+        OracleRef(
+            params.coreAddress,
+            params.oracleAddress,
+            params.backupOracle,
+            params.decimalsNormalizer,
+            params.doInvert
+        )
+    {
+        underlyingToken = _underlyingToken;
+
+        _setContractAdminRole(TribeRoles.PSM_ADMIN_ROLE);
+    }
+
+    /// @notice withdraw assets from PSM to an external address
+    function withdraw(address to, uint256 amount)
+        external
+        virtual
+        override
+        onlyPCVController
+    {
+        _withdrawERC20(address(underlyingToken), to, amount);
+    }
+
+    /// @notice function to redeem VOLT for an underlying asset
+    function redeem(
+        address to,
+        uint256 amountVoltIn,
+        uint256 minAmountOut
+    ) external virtual override whenNotPaused returns (uint256 amountOut) {
+        amountOut = getRedeemAmountOut(amountVoltIn);
+        require(
+            amountOut >= minAmountOut,
+            "PegStabilityModule: Redeem not enough out"
+        );
+
+        _beforeVoltRedeem(to, amountVoltIn, minAmountOut);
+
+        IERC20(volt()).safeTransferFrom(
+            msg.sender,
+            address(this),
+            amountVoltIn
+        );
+
+        underlyingToken.safeTransfer(to, amountOut);
+
+        emit Redeem(to, amountVoltIn, amountOut);
+
+        _afterVoltRedeem(to, amountVoltIn, minAmountOut);
+    }
+
+    /// @notice function to buy VOLT for an underlying asset
+    function mint(
+        address to,
+        uint256 amountIn,
+        uint256 minAmountVoltOut
+    ) external virtual override whenNotPaused returns (uint256 amountVoltOut) {
+        amountVoltOut = getMintAmountOut(amountIn);
+
+        require(
+            amountVoltOut >= minAmountVoltOut,
+            "PegStabilityModule: Mint not enough out"
+        );
+        require(
+            volt().balanceOf(address(this)) >= amountVoltOut,
```

Comment:
this check adds additional gas overhead in the happy path when the tx would have succeeded. I think we should remove this check

Reply:
this makes sense we can just rely on the ERC20: transfer amount exceeds balance